langchain/prompts/few_shot.py

```diff
@@ -120,8 +140,73 @@ def _prompt_type(self) -> str:
         """Return the prompt type key."""
         return "few_shot"
 
-    def dict(self, **kwargs: Any) -> Dict:
-        """Return a dictionary of the prompt."""
-        if self.example_selector:
-            raise ValueError("Saving an example selector is not currently supported")
-        return super().dict(**kwargs)
+
+# For backwards compatibility.
+FewShotPromptTemplate = FewShotStringPromptTemplate
+
+
+class FewShotChatPromptTemplate(BaseChatPromptTemplate, BaseFewShotPromptTemplate):
+    """Prompt template that contains few shot examples."""
+
+    example_input: str
+    example_output: str
+    examples: List[Dict[str, str]]
+
+    @property
+    def _prompt_type(self) -> str:
+        """Return the prompt type key."""
+        return "few_shot_chat"
+
+    def format(self, **kwargs: Any) -> Sequence[BaseMessage]:
+        """Format to a sequence of BaseMessages."""
+
+        message_prompts = []
+        if self.prefix:
+            prefix_message_prompt = SystemMessagePromptTemplate(
+                prompt=StringPromptTemplate.from_template(self.prefix)
+            )
+            message_prompts.append(prefix_message_prompt)
+
+        # TODO: add support for example selectors
+        for example in self._get_examples():
+            message_prompts.append(
+                HumanMessagePromptTemplate(
+                    prompt=StringPromptTemplate.from_template(
+                        self.example_input.format(**example)
+                    )
+                ),
+            )
+            message_prompts.append(
+                AIMessagePromptTemplate(
+                    prompt=StringPromptTemplate.from_template(
+                        self.example_output.format(**example)
+                    )
+                ),
+            )
+
+        # construct the suffix message
+        suffix_message_prompt = HumanMessagePromptTemplate(
+            prompt=StringPromptTemplate.from_template(self.suffix)
+        )
+        message_prompts.append(suffix_message_prompt)
+
+        chat_prompt_template = ChatPromptTemplate.from_messages(message_prompts)
+        return chat_prompt_template.format(**kwargs)
+
+
+if __name__ == "__main__":
+    few_shot = FewShotChatPromptTemplate(
+        prefix="You are a helpful assistant. You are helping translate from {source_language} to {target_language}.",
+        suffix="{text}",
+        example_input="{input}",
+        example_output="{output}",
+        examples=[
+            {"input": "Hello", "output": "Bonjour"},
+            {"input": "Goodbye", "output": "Au revoir"},
+            {"input": "Thank you", "output": "Merci"},
+            {"input": "I am sorry", "output": "Je suis désolé"},
+        ],
+        input_variables=["source_language", "target_language", "text"],
```

Comment:
Just tricky, since examples themselves are conditional on source/target language in this case. I think in general, it'll be hard to "tweak the parameters of the prompt" without also tweaking the examples themselves. Not sure what that means for the usefulness of FewShotChatPromptTemplate as a template though, beyond inserting text into the final `HumanMessage`.

Separately, could probably also define sensible defaults for example_input/output, given most examples don't need to be wrapped with anything extra there. (And increases the utility of the final answer, since the final text "suffix" isn't likely to be wrapped in the same way.

Suffix also feels weird, since that's where the importance of the _actual question being asked_ goes, right? Is that just legacy of what it was called historically? Wondering if there's a better name for "the actual data input this prompt attempts to answer for you"

Reply:
Yep, definitely. I think this is where example selectors can help, they can choose examples based on the parameters of the prompt.

Good point about the more sensible defaults for example_input/output.

Suffix is mainly named the way it is for legacy purposes, can change it for `FewShotChatPromptTemplate` but a change wouldn't be bw compat for the string version